walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,96 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.ListIterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Initiatize return data structure and set duration parameter
+    Collection<TimeRange> availTimes = new HashSet<>();
+    ArrayList<TimeRange> bookTimes = new ArrayList<TimeRange>();
+    final int length = (int) request.getDuration();
+    Collection<String>  requestAttendees = request.getAttendees();
+    
+    // Create a List of relevant events that will restrict available time ranges
+    bookTimes = findTimeConflicts(requestAttendees, events);
+
+    // Check that there are booked meetings before sorting it and running the available function
+    if (bookTimes.isEmpty()) {
+            availTimes.add(bookTimes.get(0).WHOLE_DAY);
+    }
+
+    else {
+        Collections.sort(bookTimes, bookTimes.get(0).ORDER_BY_START);
+        availTimes = availableTimes(bookTimes, length);
+    }
+
+    return availTimes;
+  }
+
+  public ArrayList<TimeRange> findTimeConflicts(Collection<String> requestAttendees, Collection<Event> events){
+    // Create a List of relevant events that will restrict available time ranges
+    ArrayList<TimeRange> bookTimes = new ArrayList<TimeRange>();
+
+    for (Iterator<Event> eventIterator = events.iterator(); eventIterator.hasNext();){
+        Event event = eventIterator.next();
+      
+        for (Iterator<String> attendIterator = requestAttendees.iterator(); attendIterator.hasNext();){
+            String attendee = attendIterator.next();
+           
+            if(event.getAttendees().contains(attendee) && !bookTimes.contains(event.getWhen())){
+                bookTimes.add(event.getWhen());
+            }
+        }
+    }
+    return bookTimes;
+  }
+
+  public Collection<TimeRange> availableTimes(ArrayList<TimeRange> bookTimes, int length){
+    Collection<TimeRange> availTimes = new HashSet<>();
+    if(length > bookTimes.get(0).END_OF_DAY) {
+        return availTimes;
+    }
+// My problems are in the for loop.I believe that my conditionals are inefficient and I'm still working on it.
+    else {
+    
+        for(ListIterator<TimeRange> bookIterator = bookTimes.listIterator(); bookIterator.hasNext();) {
+            TimeRange bookedTime = bookIterator.next();
+            
+            if(bookIterator.hasPrevious()) {
+                TimeRange prevBookedTime = bookIterator.previous();
+                if(!bookedTime.overlaps(prevBookedTime) && (prevBookedTime.end() + length <= bookedTime.start())) {
+                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.end(), bookedTime.start(), false));
+                }
+
```

Comment:
A good way to format if/else is like this:

if (cond) {

} else if (cond) {

} else {

}

That way the formatting gives an extra clue that you're looking at an else/if condition.

Reply:
I tried this formatting in my recent commits. Was it done correctly?